File: algorithms/2_medium/M/01007_minimum_domino_rotations_for_equal_row/solution.c
#define INF INT_MAX

inline static size_t calc(const int val, const int len, const int *tops, const int *bottoms) {
    int need_top = 0;
    int need_bottom = 0;
    for (int i = 0; i < len; i += 1) {
        if (tops[i] != val && bottoms[i] != val) {
            return INF;
        }

        if (tops[i] != val) {
            need_top += 1;
        }

        if (bottoms[i] != val) {
            need_bottom += 1;
        }
    }

    return MIN(need_top, need_bottom);
}

int minDominoRotations(int* tops, int tops_len, int* bottoms, int bottoms_len) {
    assert(tops_len > 0 && tops_len == bottoms_len && tops != NULL && bottoms != NULL);

    const int cost_top_first = calc(tops[0], tops_len, tops, bottoms);
    const int cost_bottom_first = calc(bottoms[0], tops_len, tops, bottoms);

    if (cost_top_first == INF && cost_bottom_first == INF) {
        return -1;
    }

    return MIN(cost_top_first, cost_bottom_first);
}

File: algorithms/2_medium/M/01007_minimum_domino_rotations_for_equal_row/solution.rs
impl Solution {
    #[inline(always)]
    fn calc(val: i32, tops: &[i32], bottoms: &[i32]) -> Option<usize> {
        let len = tops.len();
        let (mut need_top, mut need_bottom) = (0_usize, 0_usize);
        for i in 0..len {
            if tops[i] != val && bottoms[i] != val {
                return None;
            }

            if tops[i] != val {
                need_top += 1;
            }

            if bottoms[i] != val {
                need_bottom += 1;
            }
        }

        Some(need_top.min(need_bottom))
    }

    pub fn min_domino_rotations(tops: Vec<i32>, bottoms: Vec<i32>) -> i32 {
        assert!(!tops.is_empty() && tops.len() == bottoms.len());

        let cost_top_first = Self::calc(tops[0], &tops, &bottoms);
        let cost_bottom_first = Self::calc(bottoms[0], &tops, &bottoms);

        match (cost_top_first, cost_bottom_first) {
            (None, None) => -1,
            (Some(val1), None) => val1.try_into().unwrap(),
            (None, Some(val2)) => val2.try_into().unwrap(),
            (Some(val1), Some(val2)) => val1.min(val2).try_into().unwrap(),
        }
    }
}
